MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,2008 @@
+    for (size_t i = 0, ie = vertices->size(); i < ie; ++i) {
+      reco::Vertex& v = vertices->at(i);
+      const int ntracks = v.nTracks();
+      const double vchi2 = v.normalizedChi2();
+      const double vndof = v.ndof();
+      const double vx = v.position().x() - bsx;
+      const double vy = v.position().y() - bsy;
+      const double vz = v.position().z() - bsz;
+      const double rho = mag(vx, vy);
+      const double phi = atan2(vy, vx);
+      const double r = mag(vx, vy, vz);
+      for (const auto& r : vertex_track_set(v)) {
+        if (track_use.find(r) != track_use.end())
+          track_use[r] += 1;
+        else
+          track_use[r] = 1;
+      }
+      h_output_afterdzfit_vertex_ntracks->Fill(ntracks);
+      vertex_ntracks.push_back(ntracks);     
+      if (ntracks >= 5) {
+        count_5trk_vertices++;
+        Measurement1D dBV_Meas1D = vertex_dist_2d.distance(v, fake_bs_vtx);
+        double dBV = dBV_Meas1D.value();
+        double bs2derr = dBV_Meas1D.error();
+
+        if (vchi2 < 5 && ntracks >= 5 && bs2derr < 0.0025) {
+          h_output_afterdzfit_vertex_nm1_bsbs2ddist->Fill(dBV);
+        }
+        if (vchi2 < 5 && ntracks >= 5 && dBV > 0.01) {
+          h_output_afterdzfit_vertex_nm1_bs2derr->Fill(bs2derr);
+        }
+
+      }
+
+      for (auto it = v.tracks_begin(), ite = v.tracks_end(); it != ite; ++it) {
+        h_output_afterdzfit_vertex_track_weights->Fill(v.trackWeight(*it));
+
+        reco::TransientTrack seed_track;
+        seed_track = tt_builder->build(*it.operator*());
+        std::pair<bool, Measurement1D> tk_vtx_dist = track_dist(seed_track, v);
+        h_output_afterdzfit_vertex_tkvtxdist->Fill(tk_vtx_dist.second.value());
+        h_output_afterdzfit_vertex_tkvtxdisterr->Fill(tk_vtx_dist.second.error());
+        h_output_afterdzfit_vertex_tkvtxdistsig->Fill(tk_vtx_dist.second.significance());
+      }
+
+      h_output_afterdzfit_vertex_chi2->Fill(vchi2);
+      h_output_afterdzfit_vertex_ndof->Fill(vndof);
+      h_output_afterdzfit_vertex_x->Fill(vx);
+      h_output_afterdzfit_vertex_y->Fill(vy);
+      h_output_afterdzfit_vertex_rho->Fill(rho);
+      h_output_afterdzfit_vertex_phi->Fill(phi);
+      h_output_afterdzfit_vertex_z->Fill(vz);
+      h_output_afterdzfit_vertex_r->Fill(r);
+
+      for (size_t j = i + 1, je = vertices->size(); j < je; ++j) {
+        const reco::Vertex& vj = vertices->at(j);
+        const double vjx = vj.position().x() - bsx;
+        const double vjy = vj.position().y() - bsy;
+        const double phij = atan2(vjy, vjx);
+        Measurement1D v_dist = vertex_dist(vj, v);
+        h_output_afterdzfit_vertex_paird2d->Fill(mag(vx - vjx, vy - vjy));
+        h_output_afterdzfit_vertex_paird2dsig->Fill(v_dist.significance());
+        h_output_afterdzfit_vertex_pairdphi->Fill(reco::deltaPhi(phi, phij));
+      }
+
+
+    }
+
+    h_n_at_least_5trk_output_afterdzfit_vertices->Fill(count_5trk_vertices);
+
+  }
+

```

Comment:
Am I correct in thinking that the plots filled for each of your histos_output levels are the same? i.e. that the code snippet highlighted here and that in https://github.com/DisplacedVertices/cmssw-usercode/pull/5/files?w=1#diff-eaef89c993d668ab99b433c8e0844dbd1b7a8260a673b2d83468451a634fa1caR1506-R1581 are identical, and the same is true for https://github.com/DisplacedVertices/cmssw-usercode/pull/5/files?w=1#diff-eaef89c993d668ab99b433c8e0844dbd1b7a8260a673b2d83468451a634fa1caR1787-R1860?
 (or perhaps identical other than `h_output_aftersharedjets_n_onetracks`, which in principle could be elsewhere)
 
If so, we might be able to make a function that does all of the plot filling. It would have to take the vertex and beamspot as input, but that might be it? Well, and the "output" type, so that the code knows which plot to fill.

Let me know if that seems accurate to you. I could try to play around with it once you confirm that these snippets are identical (other than the single plot mentioned, and the histogram names). This improvement would be to try to avoid significant code duplication.

Reply:
Yes, that is the idea that all similar plots are made after vertices are changed. I agree with you to proceed with this development.

Reply:
As mentioned on Slack: done in https://github.com/DisplacedVertices/cmssw-usercode/commit/c7e18622d9dd7ed5b51ea31fa16ffab25e709090. We can adjust if desired, but a test job (preferably w/ signal) would be useful to ensure that none of the plots changed.

Reply:
Done.